Add CQL forwarding for strongly consistent tables

wmitros opened the pull request:
In this series we add support for forwarding strongly consistent CQL requests to suitable replicas, so that clients can issue reads/writes to any node and have the request executed on an appropriate tablet replica (and, for writes, on the Raft leader). We return the same CQL response as what the user would get while sending the request to the correct replica and we perform the same logging/stats updates on the request coordinator as if the coordinator was the appropriate replica.

The core mechanism of forwarding a strongly consistent request is sending an RPC containing the user's cql request frame to the appropriate replica and returning back a ready, serialized `cql_transport::response`. We do this in the CQL server - it is most prepared for handling these types and forwarding a request containing a CQL frame allows us to reuse near-top-level methods for CQL request handling in the new RPC handler (such as the general `process`)

For sending the RPC, the CQL server needs to obtain the information about who should it forward the request to. This requires knowledge about the tablet raft group members and leader. We obtain this information during the execution of a `cql3/strong_consistency` statement, and we return this information back to the CQL server using the generalized `bounce_to_shard` `response_message`, where we now store the information about either a shard, or a specific replica to which we should forward to. Similarly to `bounce_to_shard`, we need to handle this `result_message` in a loop - a replica may move during statement execution, or the Raft leader can change. We also use it for forwarding strongly consistent writes when we're not a member of the affected tablet raft group - in that case we need to forward the statement twice - once to any replica of the affected tablet, then that replica can find the leader and return this information to the coordinator, which allows the second request to be directed to the leader.

This feature also allows passing through exception messages which happened on the target replica while executing the statement. For that, many methods of the `cql_transport::cql_server::connection` for creating error responses needed to be moved to `cql_transport::cql_server`. And for final exception handling on the coordinator, we added additional error info to the RPC response, so that the handling can be performed without having the `result_message::exception` or `exception_ptr` itself.

Fixes [SCYLLADB-71](https://scylladb.atlassian.net/browse/SCYLLADB-71)

[SCYLLADB-71]: https://scylladb.atlassian.net/browse/SCYLLADB-71?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

gleb-cloudius commented on cql3/result_set.cc:
```diff
@@ -188,4 +191,80 @@ make_empty_metadata() {
     return empty_metadata_cache;
 }
 
+result_set_serialized serialize_result_set(const result_set& rs) {
```
It does not looks like this serializes anything. It creates serializable structure. And BTW do we need this intermediate serializable structure? Is it absolutely impossible to define result_set in the idl?

gleb-cloudius commented on exceptions/coordinator_result.hh:
```diff
@@ -43,4 +44,62 @@ using coordinator_result = bo::result<T,
     utils::exception_container_throw_policy
 >;
 
+// Serializable representation of mutation_write_timeout_exception for RPC
```
But all cql message are already serialized, just not by our rpc, but by cql server code. Why not reuse the format/code?

gleb-cloudius replied:
> we only have the serialization in the server code, the deserialization needs to be added from scratch.

Do we need de-serialization at all? We need to pass it to a client as is, no?

gleb-cloudius replied:
I would guess we will need new `forward_execute` method.

wmitros commented:
> What is CQL forwarding?

We also referred to is as proxying. If the raft group leader of a tablet affected by a request is different than the request coordinator, we'll use the forwarding to send the request from the coordinator to the leader. For most requests we won't use it, because the drivers are supposed to be leader-aware, but for the cases where the request is received on the wrong node, we'll use this mechanism instead of just rejecting the request.

gleb-cloudius commented on idl/forward_cql.idl.hh:
```diff
@@ -0,0 +1,66 @@
+};
+
+struct query_tablet_leader_response {
+    std::optional<utils::UUID> leader_host_id;
```
UUID support `operator!` as well.

gleb-cloudius commented on service/forward_cql_service.cc:
```diff
@@ -0,0 +1,574 @@
+                    .leader_host_id = utils::UUID{}
+                };
+            }
+        }, raft_server.begin_mutate());
```
`begin_mutate` does not look like a good name now, no?

piodul commented on idl/paging_state.idl.hh:
```diff
@@ -22,19 +23,6 @@ enum class read_repair_decision : uint8_t {
-    clustered,
-    partition_end,
-};
-
```
> don't redefine bound_weight and partition_region in paging_state.idl.hh 

Could you send this as a separate, single-commit PR? This looks like an improvement that is independent from the rest of the PR.

wmitros replied:
Ok, but I'll still need to base this PR on it

piodul commented on idl/forward_cql.idl.hh:
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2025-present ScyllaDB
```
Nit: happy new year, it's 2026 now.

wmitros commented on service/forward_cql_service.cc:
```diff
@@ -0,0 +1,574 @@
+                .leader_host_id = std::nullopt
+            };
+        }
+        auto raft_server = co_await gm.acquire_server(group_id);
```
I added a check for this in the rebase

wmitros commented on service/forward_cql_service.cc:
```diff
@@ -0,0 +1,574 @@
+                    return locator::host_id{};
+                }
+            }, raft_server.begin_mutate());
+
```
I extracted the code to a separate method in the rebase

wmitros commented on test/cluster/test_forward_cql.py:
```diff
@@ -0,0 +1,171 @@
+
+logger = logging.getLogger(__name__)
+
+
```
I started developing this when `test_strong_consistency.py` was not present, but now it makes most sense to put these tests together - I moved them in the rebase

wmitros commented on service/forward_cql_service.cc:
```diff
@@ -0,0 +1,574 @@
+    auto& tab = get_column_family(*stmt);
+
+    if (is_stale_topology(tab, topology_version)) {
+        co_return forward_cql_execute_response{
```
It was part of the design doc, but if we're checking on the target node if we're a leader and we return only after committing the change, it's probably not needed anymore - I removed it in the rebase

wmitros commented:
I split the patch into commits and addressed most of the comments, but I didn't fix some of the most recent ones yet and I still have to add a cover letter
In the rebase I added exception passthrough and mostly reorganized the code in the forward_cql_service. I also rebased on top of the newest https://github.com/scylladb/scylladb/pull/27614

gusev-p commented:
> Why can't we reuse the same pattern we're using for shard bouncing -- try to execute locally, get some bounce response message, extract where to bounce from it, do the forwarding?

One difference is that the target node can be unavailable and we need some way to decide what to do in that case. I'm not sure we need a complicated retry strategies here though. The `strong_consistency/coordinator` may consult `_raft_gr.failure_detector()`, choose a currently alive replica to redirect to, and then we fail the current request if this replica turned out to be unavailable -- the driver should retry in that case.

One way is to generalize `class result_message::bounce_to_shard` to something like:
```
class result_message::bounce : public result_message {
public:
  struct bounce_to_shard {
    unsigned shard;
    cql3::computed_function_values cached_fn_calls
  };
  struct bounce_to_node {
    host_id host;
    shard_id shard;
  };
  using bounce_target = std::variant<bounce_to_shard, bounce_to_node>;
  bounce(bounce_target target);
};
```

wmitros commented:
> > Why can't we reuse the same pattern we're using for shard bouncing -- try to execute locally, get some bounce response message, extract where to bounce from it, do the forwarding?
> 
> One difference is that the target node can be unavailable and we need some way to decide what to do in that case. I'm not sure we need a complicated retry strategies here though. The `strong_consistency/coordinator` may consult `_raft_gr.failure_detector()`, choose a currently alive replica to redirect to, and then we fail the current request if this replica turned out to be unavailable -- the driver should retry in that case.

Thanks, that's one thing I wasn't sure about

> One way is to generalize `class result_message::bounce_to_shard` to something like:
> 
> ```
> class result_message::bounce : public result_message {
> public:
>   struct bounce_to_shard {
>     unsigned shard;
>     cql3::computed_function_values cached_fn_calls
>   };
>   struct bounce_to_node {
>     host_id host;
>     shard_id shard;
>   };
>   using bounce_target = std::variant<bounce_to_shard, bounce_to_node>;
>   bounce(bounce_target target);
> };
> ```

Looks good. Do you think it's worthwhile to provide the shard when bouncing to another node? We'll need to have the shard bouncing retry loop after forwarding either way

gusev-p commented:
> Looks good. Do you think it's worthwhile to provide the shard when bouncing to another node? We'll need to have the shard bouncing retry loop after forwarding either way

Currently there is some magic in the seastar sockets that tries to ensure that a shard on one node connects to the same shard on another node (grep by `server_socket::load_balancing_algorithm`, `posix_server_socket_impl::accept`,  `posix_socket_impl::find_port_and_connect`). On the other hand, the `posix_server_socket_impl::accept` just does the regular `smp::submit_to` if we ended up on the wrong shard. Also, the number of shards could be different on different nodes (affects the choice of the client port), so this magic is not guaranteed to work.

I think we can pass the target shard to `result_message::bounce` and do the submit_to on the target node, so that not to waste the first attempt. The tablet can migrate to another shard (or node) before that (we'll retry in that case), but this should be rare.

wmitros commented on service/forward_cql_service.cc:
```diff
@@ -0,0 +1,574 @@
+    const auto this_host = erm->get_token_metadata().get_my_id();
+
+    // Check if we're a replica of this tablet
+    if (locator::contains(tablet_info.replicas, this_host)) {
```
I removed the replica selection from the `forward_cql_service` in the last rebase

wmitros commented on service/forward_cql_service.cc:
```diff
@@ -0,0 +1,574 @@
+    }
+    std::optional<std::unique_ptr<cql_transport::response>> response;
+    if (result->move_to_shard()) {
+        auto target_shard = result->move_to_shard().value();
```
Bouncing to another shard or another node is now in one loop. We also handle both cases in one place on the receiving node

gusev-p commented on service/forward_cql_service.cc:
```diff
@@ -284,13 +284,22 @@ forward_cql_service::forward_cql_without_checking_exceptions(
+        } else {
+            auto target = (*bounce_msg)->move_to_node();
+            const auto my_host_id = _qp.db().real_database().get_token_metadata().get_topology().my_host_id();
+            throw exceptions::invalid_request_exception(format(
```
> Of course, it requires careful usage of try_future everywhere.

It's unfortunate that unwinding is the default, not the `try_future`-style forwarding. In some alternate universe we could go over the Scylla codebase, convert all try/catch-es around co_await to `coroutine::as_future`, then build seastar with some compile flag that forces `co_await some_func()` to forward exceptions instead of unwinding them.

wmitros commented on transport/response.hh:
```diff
@@ -88,6 +88,24 @@ public:
+    }
+
+    // Extract the body as bytes (linearizes if needed)
+    bytes extract_body() {
```
I think I managed to change it so we perform only one allocation and copy

wmitros commented on cql3/statements/prepared_statement.hh:
```diff
@@ -50,6 +50,7 @@ public:
     const std::vector<seastar::lw_shared_ptr<column_specification>> bound_names;
     const std::vector<uint16_t> partition_key_bind_indices;
     std::vector<sstring> warnings;
+    bool requires_forwarding = false;  // Set if keyspace is strongly consistent
```
I removed this bool in the rebase

wmitros commented on transport/server.cc:
```diff
@@ -1282,9 +1283,43 @@ cql_server::connection::process(uint16_t stream, request_reader in, service::cli
+    // Handle error logging and stats if this was an error response
+    if (result.error_info) {
+        clogger.debug("{}: {}", _client_state.get_remote_address(), result.error_info->log_message);
+        try { ++_server._stats.errors[static_cast<exceptions::exception_code>(result.error_info->exception_code)]; } catch(...) {}
```
Ok

piodul commented:
> > @wmitros Question: how easy it will be for Alternator to use forwarding that is implemented on the CQL level?
> 
> Hard to tell - AFAIK Alternator doesn't use statements which are needed for strongly-consistent tables right now.
> 
> Though for sure a lot would be needed to be added - I don't think Alternator uses the cql server. But for example, if we went with the approach currently present in the PR, it would still be problematic because here we rely on prepared statements in forwarding requests which we wouldn't have in Alternator

Just leaving a comment here - we discussed this on Slack with others in the Cluster team and it looks like the most sensible option would be for Alternator to implement some kind of forwarding as well. The Alternator protocol seems to be stateless and even simpler than CQL.

An alternative to CQL/Alternator-level forwarding would be to forward on the strong consistency level (i.e. introduce some RPCs that `strong_consistency::coordinator` would use) but we are not pursuing this direction as maintaining CQL-level forwarding seems less tedious in the long term.

piodul commented on idl/forward_cql.idl.hh:
```diff
@@ -0,0 +1,75 @@
+    // Protocol version for parsing
+    uint8_t cql_version;
+    // Stream ID for response
+    uint16_t stream;
```
Does the receiving side need the stream ID for anything? Stream ID is only relevant for direct client-server communication. It's sufficient that only the node that talks directly to the driver knows the stream ID.

piodul reviewed (commented):
Leaving another batch of comments... There is a lot to take in here, so I will need another (if not more) sitting.

wmitros commented on transport/server.cc:
```diff
@@ -432,7 +410,55 @@ future<> cql_server::unregister_forwarding_handlers() {
+
+    if (!final_result) {
+        co_return co_await final_result.assume_error().as_exception_future<forward_cql_execute_response>();
+    }
```
I'm not sure how I can use it here - do you mean to wrap the expression in it?

gusev-p commented on transport/server.cc:
```diff
@@ -1462,6 +1817,92 @@ process_batch_internal(service::client_state& client_state, sharded<cql3::query_
+            auto request_bytes = linearize_istream(is);
+            auto q_state = std::make_unique<cql_query_state>(client_state, trace_state, permit);
+            auto rpc_timeout = client_state.get_timeout_config().read_timeout;
+            auto deadline = seastar::lowres_clock::now() + rpc_timeout;
```
Why we need this timeout? And why we use read_timeout, I guess for some requests we should be using `write_timeout` (or some other timeout).

gusev-p commented on transport/server.cc:
```diff
@@ -342,6 +349,253 @@ cql_server::cql_server(sharded<cql3::query_processor>& qp, auth::service& auth_s
+    }
+}
+
+future<bytes> cql_server::send_forward_prepare(
```
why we have this method? Used only once, we doesn't have similar wrapper for `send_forward_cql_execute`. Seems simpler to inline it.

gusev-p commented on transport/server.hh:
```diff
@@ -219,14 +268,16 @@ public:
     using result_with_foreign_response_ptr = exceptions::coordinator_result<foreign_ptr<std::unique_ptr<cql_server::response>>>;
-    using result_with_bounce_to_shard = foreign_ptr<seastar::shared_ptr<messages::result_message::bounce_to_shard>>;
-    using process_fn_return_type = std::variant<result_with_foreign_response_ptr, result_with_bounce_to_shard>;
+    using result_with_bounce = foreign_ptr<seastar::shared_ptr<messages::result_message::bounce>>;
```
After digging a bit deeper I see that:
* there is a slight difference between 'process' functions in how they deal with metadata_id. process_execute_internal extracts metadata_id and passes it to make_result, other functions behave differently.
* Forward response is returned in the form of cql_transport::response, which makes sense

I don't like these complicated types and conversions, but I don't see how this can be improved without big refactorings. Nothing actionable for this PR.

gusev-p commented on transport/server.cc:
```diff
@@ -344,6 +348,131 @@ cql_server::cql_server(sharded<cql3::query_processor>& qp, auth::service& auth_s
+                        : db::timeout_clock::duration::zero();
+                }
+                service::client_state cs(service::client_state::internal_tag{},
+                    ::timeout_config{d, d, d, d, d, d, d});
```
Also, we need to use a proper timeout_config(). The `rpc::opt_time_point` timeout parameter is not enough -- it's always read_timeout(), won't work properly with write requests.

gusev-p commented on transport/server.cc:
```diff
@@ -342,6 +349,253 @@ cql_server::cql_server(sharded<cql3::query_processor>& qp, auth::service& auth_s
+{
+    auto opcode = static_cast<cql_binary_opcode>(req.opcode);
+
+    fragmented_temporary_buffer ftb(reinterpret_cast<const char*>(req.request_buffer.data()), req.request_buffer.size());
```
This copies `request_buffer` bytes unnecessary. Can't we construct `fragmented_temporary_buffer::istream` this way:
```
temporary_buffer<char> buf {
    reinterpret_cast<const char*>(req.request_buffer.data()),
    req.request_buffer.size(),
    deleter{}
};
fragmented_temporary_buffer::istream is({std::move(buf}, temp_linearization_buffer);
```
The linearization buffer shouldn't be actually used since the bytes are already linearized.

I saw the same in other places in this PR.

gusev-p commented on transport/server.cc:
```diff
@@ -342,6 +349,253 @@ cql_server::cql_server(sharded<cql3::query_processor>& qp, auth::service& auth_s
+                service::query_state qs(cs, trace_state_ptr, empty_service_permit());
+                tracing::trace(qs.get_trace_state(), "Handling forwarded CQL request from {} on shard {}", src_host, this_shard_id());
+
+                co_return co_await shard_svc.handle_forward_execute(qs, req).then_wrapped([&qs, &req, &shard_svc] (future<forward_cql_execute_response> f) mutable {
```
And btw, `&req` in `then_wrapped` lambda only works because of an `smp::invoke_on` specific lifetime handing -- it doesn't destroy lambda until the future is resolved. In general, lambda and its closures are destroyed as soon as the synchronous function returns. With co_await we don't need to think about such nuances -- everything lives in the coroutine frame.

gusev-p commented on transport/server.cc:
```diff
@@ -1462,6 +1817,92 @@ process_batch_internal(service::client_state& client_state, sharded<cql3::query_
+}
+
+future<cql_server::process_fn_return_type>
+cql_server::process(uint16_t stream, request_reader in, service::client_state& client_state, service_permit permit,
```
I wonder if we could simplify the argument list a bit by passing along `cql_query_state` instead of `client_state + service_permit + trace_state`.

wmitros replied:
Perhaps we could but I'd prefer to avoid unnecessary changes in this PR

gusev-p commented on transport/server.cc:
```diff
@@ -342,6 +349,253 @@ cql_server::cql_server(sharded<cql3::query_processor>& qp, auth::service& auth_s
+                co_return co_await shard_svc.handle_forward_execute(qs, req).then_wrapped([&qs, &req, &shard_svc] (future<forward_cql_execute_response> f) mutable {
+                    if (f.failed()) {
+                        auto eptr = f.get_exception();
+                        if (try_catch<exceptions::prepared_query_not_found_exception>(eptr)) {
```
We can move this try/catch closer to the source of the exceptions -- it's generally easier to reason about the code when exceptions are handled as close to the place where they can be throws as possible.

gusev-p commented on cql3/statements/strong_consistency/modification_statement.cc:
```diff
@@ -62,8 +66,21 @@ future<shared_ptr<result_message>> modification_statement::execute_without_check
     using namespace service::strong_consistency;
     if (const auto* redirect = get_if<need_redirect>(&mutate_result)) {
-        co_return co_await redirect_statement(qp, options, redirect->target);
+        if (db::timeout_clock::now() >= timeout) {
```
I don't see a good reason why timeouts need to be handled in this place. In [this PR](https://github.com/scylladb/scylladb/pull/28526) we should start handling timeouts in `coordinator.mutate` itself. (/cc @dawmd)

If we want the timeouts to be checked before (and during) redirect we should do that in the general redirect/forwarding handling code.

piodul commented on transport/server.cc:
```diff
@@ -405,6 +412,34 @@ void cql_server::register_forwarding_handlers() {
+            if (req.keyspace) {
+                cs.set_raw_keyspace(*req.keyspace);
+            }
+            service::query_state qs(cs, tracing::trace_state_ptr{}, empty_service_permit());
```
I think that we should construct a proper client state as well, as discussed in this thread: https://github.com/scylladb/scylladb/pull/27517/commits/ccf9449f00bb5c6dbc93fb0d13d3f61bc22f2659#r2783582334. We should have a concept of forwarded client state which is somewhere in the middle between external and internal.

wmitros commented on transport/server.cc:
```diff
@@ -342,6 +349,253 @@ cql_server::cql_server(sharded<cql3::query_processor>& qp, auth::service& auth_s
+    cql3::computed_function_values cached_fn_calls)
+{
+    forward_cql_execute_request req{
+        .request_buffer = request_buffer,
```
I removed the copy

wmitros commented on transport/server.cc:
```diff
@@ -284,6 +289,8 @@ cql_server::cql_server(sharded<cql3::query_processor>& qp, auth::service& auth_s
         return;
     }
 
+    register_forwarding_handlers();
```
It was unfortunate that the diff cutoff the line at the top. We had:
```
    if (used_by_maintenance_socket) {
        return;
    }

    register_forwarding_handlers();
```
I looked at the `stop()` code and we do not check if we're the maintenance socket but from what I saw in the code, unregistering RPC handlers is supposed to be idempotent

wmitros commented on transport/server.cc:
```diff
@@ -503,8 +538,37 @@ cql_server::forward_cql(
+            if (!prepared) {
+                clogger.info("Prepared statement with cache key {} not found locally for prepare request, cannot prepare on target {}", cache_key.key(), current_host);
+                co_return coroutine::exception(std::make_exception_ptr(exceptions::prepared_query_not_found_exception(cache_key.key())));
+            }
```
We can get the cache key from the `prepared_not_found` response, so I used it instead of parsing in the last rebase

wmitros commented on service/strong_consistency/coordinator.cc:
```diff
@@ -36,6 +38,24 @@ struct coordinator::operation_ctx {
 };
 
+// Select closest replica from a tablet replica set, preferring replicas in same rack
+static locator::host_id select_closest_replica(const locator::tablet_replica_set& replicas, const dht::token& token, const locator::topology& topo, shared_ptr<raft::failure_detector> failure_detector) {
```
> But here we forward to one of the replicas since we are not the replica. Our raft view is irrelevant here since we are not part of the raft group.

Ok, I used the gossiper instead in the last rebase.

> Actually, I don't see why can't we just put the host_id of the leader into the `tablet_raft_info` -- leader changes should be rare. The frequency of leader changes is comparable to tablet movements. We already use an optimization in `topology_state_load` -- only changed tablets are reloaded from system tables, so tablet_map rebuilds shouldn't be a problem. We would also need the information who is the leader [to pass it to the drivers](https://github.com/scylladb/scylladb/issues/25948).

We ended up deciding to just cache the leaders instead of storing them directly in a table, but I'll do that as a follow-up

avikivity commented:
Looks good.

How long do the new tests take to run? 

Please show before/after for perf_cql_raw (smp 1).

gusev-p commented on transport/server.cc:
```diff
@@ -357,6 +377,205 @@ cql_server::cql_server(sharded<cql3::query_processor>& qp, auth::service& auth_s
+            service::query_state qs(cs, tracing::trace_state_ptr{}, empty_service_permit());
+
+            co_await _query_processor.invoke_on_all(
+                [req, &qs] (cql3::query_processor& qp) {
```
`req` redundant copy, can be captured by ref

I don't see why we pass query_string by val, but this is preexisting, let's leave as is.

wmitros commented on cql3/query_processor.cc:
```diff
@@ -506,6 +506,12 @@ query_processor::query_processor(service::storage_proxy& proxy, data_dictionary:
+                            "forwarded_requests",
+                            _cql_stats.forwarded_requests,
+                            sm::description("Counts the total number of attempts to forward strongly consistent CQL requests to other nodes."
+                                            "One request may be forwarded multiple times if the forwarding is attempted during tablet Raft group changes.")).set_skip_when_empty(),
```
Ok

gusev-p commented on service/client_state.hh:
```diff
@@ -246,6 +264,22 @@ public:
 
+    client_state(auth::service& auth_service,
+                 qos::service_level_controller* sl_controller,
+                 const forwarded_client_state& forwarded_state)
```
probably we could `std::move` the `forwarded_client_state` here

wmitros replied:
Ok

wmitros commented on transport/server.cc:
```diff
@@ -357,6 +377,205 @@ cql_server::cql_server(sharded<cql3::query_processor>& qp, auth::service& auth_s
+            _stats.requests_forwarded_prepared_not_found++;
+            cql3::prepared_cache_key_type cache_key(std::move(response.prepared_id), req.dialect);
+
+            auto prepared = _query_processor.local().get_prepared(cache_key);
```
Not sure either, @scylladb/core-frontend please review this part

wmitros replied:
Thanks, so it looks like indeed we'll suffer a bit of lower efficiency but there should be no issue with correctness and the issue should vanish as soon as all nodes update their schema.

gusev-p commented on transport/server.cc:
```diff
@@ -388,60 +392,22 @@ void cql_server::init_messaging_service() {
+                    co_return forward_cql_execute_response{
+                        .status = forward_cql_status::error,
+                        .response_body = std::move(*response.get()).extract_body(),
+                        .response_flags = response->flags(),
```
It seems `response->flags()` could be executed after `std::move(*response.get())`. Probably nothing wrong with this in practice, but better to capture the flags separately. Same in `handle_forward_execute`.

gusev-p commented on test/cluster/test_strong_consistency.py:
```diff
@@ -552,3 +621,117 @@ def host_by_host_id(host_id):
+            traced_execute = cql.execute(stmt, (0,), trace=True)
+            res = await cql.run_async(f"SELECT value FROM {table} WHERE pk = 0")
+            assert len(res) == 1 and res[0].value == 0
+            trace = traced_execute.get_query_trace()
```
nice!

gusev-p commented on service/strong_consistency/coordinator.cc:
```diff
@@ -36,6 +39,33 @@ struct coordinator::operation_ctx {
+    }
+    if (hosts.empty()) {
+        // If all replicas are down, there's no node worth forwarding to, so we return an exception
+        throw exceptions::unavailable_exception(format("All replicas for token {} are down", token), db::consistency_level::ONE, 1, 0);
```
Ideally we should have specific exceptions handling in the drivers for strong consistency -- we shouldn't be telling users something about consistency_level-s because they are irrelevant and confusing. I've added a [todo item](https://docs.google.com/document/d/1afKeeHaCkKxER7IThHkaAQlh2JWpbqhFLIQ3CzmiXhI/edit?tab=t.0) `Strongly consistent exceptions` to think about that when we come to supporting sc on the drivers side.

wmitros replied:
From what I see in `send_message_timeout`, transport errors result in `rpc::closed_error` so I added handling for it

wmitros commented:
> dtest	[dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/ci-dtest-release-with-consistent-topology-changes/729//pipeline-overview)	❌		3m	
> [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/ci-dtest-release-with-tablets/727//pipeline-overview)	❌		3m	

No tests failed, I see it was aborted after some
```
12:01:40  error saving credentials: open /home/jenkins/.docker/config.json527574867: permission denied
```
Maybe rerunning will help

And I reran tests without PGO and LTO. New baseline:
```
55827.71 tps ( 79.6 allocs/op,   0.0 logallocs/op,  24.3 tasks/op,   50688 insns/op,   33106 cycles/op,        0 errors)
56917.33 tps ( 79.5 allocs/op,   0.0 logallocs/op,  24.3 tasks/op,   50672 insns/op,   32985 cycles/op,        0 errors)
53572.25 tps ( 79.9 allocs/op,   0.0 logallocs/op,  24.5 tasks/op,   50791 insns/op,   34556 cycles/op,        0 errors)
47196.00 tps ( 80.2 allocs/op,   0.0 logallocs/op,  24.8 tasks/op,   50862 insns/op,   38697 cycles/op,        0 errors)
55753.19 tps ( 79.7 allocs/op,   0.0 logallocs/op,  24.4 tasks/op,   50720 insns/op,   33443 cycles/op,        0 errors)
throughput:
        mean=   53853.30 standard-deviation=3914.55
        median= 55753.19 median-absolute-deviation=1974.41
        maximum=56917.33 minimum=47196.00
instructions_per_op:
        mean=   50746.54 standard-deviation=79.09
        median= 50720.25 median-absolute-deviation=58.64
        maximum=50862.00 minimum=50671.74
cpu_cycles_per_op:
        mean=   34557.47 standard-deviation=2395.80
        median= 33442.67 median-absolute-deviation=1451.48
        maximum=38697.35 minimum=32985.46
```
and with the patch:
```
49764.45 tps ( 80.1 allocs/op,   0.0 logallocs/op,  24.7 tasks/op,   50910 insns/op,   37093 cycles/op,        0 errors)
50555.23 tps ( 79.9 allocs/op,   0.0 logallocs/op,  24.6 tasks/op,   50845 insns/op,   37108 cycles/op,        0 errors)
49817.99 tps ( 80.2 allocs/op,   0.0 logallocs/op,  24.8 tasks/op,   50932 insns/op,   37267 cycles/op,        0 errors)
50661.53 tps ( 80.0 allocs/op,   0.0 logallocs/op,  24.6 tasks/op,   50858 insns/op,   36958 cycles/op,        0 errors)
51489.83 tps ( 79.9 allocs/op,   0.0 logallocs/op,  24.6 tasks/op,   50832 insns/op,   36535 cycles/op,        0 errors)
throughput:
        mean=   50457.81 standard-deviation=708.20
        median= 50555.23 median-absolute-deviation=639.82
        maximum=51489.83 minimum=49764.45
instructions_per_op:
        mean=   50875.22 standard-deviation=43.50
        median= 50857.58 median-absolute-deviation=34.65
        maximum=50932.17 minimum=50831.87
cpu_cycles_per_op:
        mean=   36992.12 standard-deviation=278.13
        median= 37092.79 median-absolute-deviation=116.21
        maximum=37266.98 minimum=36534.89

```
Still 100-200 insns/op more
